manager/dispatcher/dispatcher.go

```diff
@@ -1091,13 +1114,33 @@ func (d *Dispatcher) Assignments(r *api.AssignmentsRequest, stream api.Dispatche
 	}
 }
 
-func (d *Dispatcher) nodeRemove(id string, status api.NodeStatus) error {
+// markNodeNotReady sets the node state to some state other than READY
+func (d *Dispatcher) markNodeNotReady(id string, state api.NodeStatus_State, message string) error {
 	if err := d.isRunningLocked(); err != nil {
 		return err
 	}
 
+	// lookup the node so we can get its old address
+	// TODO(dperny) is there a faster/better way to do this?
```

Comment:
Yup, this can be simplified. What I would do is change `processUpdates` to only apply changes that exist in the `NodeStatus` structure instead of applying the whole thing atomically.

Right now it has:

```go
                                if nodeUpdate.status != nil {
                                        node.Status = *nodeUpdate.status
                                }
```

You could change this to:

```go
                                if nodeUpdate.status != nil {
                                        node.Status.State = nodeUpdate.status.State
                                        node.Status.Message = nodeUpdate.status.Message
                                        if nodeUpdate.status.Address != "" {
                                                node.Status.Address = nodeUpdate.status.Address
                                        }
                                }
```